src/sources/host_metrics/cgroups.rs

```diff
@@ -0,0 +1,271 @@
+use super::{filter_result_sync, FilterList, HostMetricsConfig};
+use crate::event::metric::Metric;
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use shared::btreemap;
+use std::fs::{self, File};
+use std::io::{self, Read};
+use std::num::ParseIntError;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+#[derive(Clone, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(default)]
+pub(super) struct CgroupsConfig {
+    #[derivative(Default(value = "3"))]
+    levels: usize,
+    base: Option<PathBuf>,
+    groups: FilterList,
+}
+
+impl HostMetricsConfig {
+    pub async fn cgroups_metrics(&self) -> Vec<Metric> {
+        let now = Utc::now();
+        CGroup::root(self.cgroups.base.as_deref())
+            .map(|root| self.recurse_cgroup(now, root, 1))
+            .unwrap_or_else(Vec::new)
+    }
+
+    fn recurse_cgroup(&self, now: DateTime<Utc>, cgroup: CGroup, level: usize) -> Vec<Metric> {
+        let mut result = Vec::new();
+
+        let tags = btreemap! {
+            "cgroup" => cgroup.name.to_string_lossy(),
```

Comment:
This makes me wonder if adding a `cgroup` tag to the existing metrics is enough to satisfy the requirements? In other words, dropping the `cgroups_*` metrics. Is that possible, or is there any reason you didn't consider this?

Reply:
I am unclear what you are asking here. Is this a question about the naming of the new metrics or of procedure in how the metrics are being generated?

Reply:
Procedure. Basically, I view cgroup information as a piece of context to the existing host-level metrics. If possible, it should be a facet to slice the existing metrics across, not an entirely new set of metrics. Does that make sense?

Reply:
It does, somewhat. We could potentially replace some of the other metrics gathering procedures with this cgroups module, but it only applies to some of the metrics (ex filesystem space, load average, and network usage are not present), only works on Linux (as opposed to our current cross-platform support), and only when v2 or hybrid cgroups are enabled (there are still some v1-only setups). I am not sure the reported numbers are identical either, though they should be close at least.

Reply:
Got it. It sounds like it would be somewhat awkward to try and accomplish that, right? Mostly because we rely on a library to collect those metrics.

Reply:
Yes, it would be awkward, at best. More so because we are trying to be as cross-platform as possible.